### ISSUE TITLE
Library gets stuck when trying to geocode an address with external key

### ISSUE THREAD

**Issue by Ismamai** [googlemaps/google-maps-services-java#161]
I am using v0.12 and trying to geocode an address with external key and the library does not return any result. In my code I create a Future that will wait for the library to return a result.

```
        request.setCallback(new PendingResult.Callback<GeocodingResult[]>() {
            @Override
            public void onResult(GeocodingResult[] result) {
                geocodeFuture.setResult(getResultFromResponse(result));
            }
```

....
I first throw the requests and then I try to get the response from the futures with a timeout of 60 seconds

geocodeFuture.get(TIMEOUT, TimeUnit.SECONDS);

I am not sure why, but it seems that the library gets stuck and it does not return any response. But if I try the url that is sent to the google server, I get the response straight away.

Is there something that I am missing ?

**Comment by Ismamai** [comment]
I was able to get a thread dump and it is stuck on the retry method. The strange thing is that if I try the url in the browser the response comes back straight away. Also I set the QPS to 50 as the limit allows. 

"OkHttp https://maps.googleapis.com/maps/api/geocode/json?client=CLIENT&address=Traxx+Tyres+Limited%2C+Maxx+House%2CWestern+Road%2CTraxx+Tyres+Limited%2C+Maxx+House%2CWestern+Road%2CBracknell%2CBerkshire%2CRG12+1Q%2CUnited+Kingdom&channel=-test-ENTERPRISE&signature=VAo8_hbLuKNdVgAqBDx-qNPDzns=" #1125 daemon prio=3 os_prio=0 tid=0x00007f3368002000 nid=0x4c66 waiting on condition [0x00007f3378163000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000006696fa410> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.ArrayBlockingQueue.take(ArrayBlockingQueue.java:403)
	at com.google.maps.internal.OkHttpPendingResult.await(OkHttpPendingResult.java:165)
	at com.google.maps.internal.OkHttpPendingResult.retry(OkHttpPendingResult.java:297)
	at com.google.maps.internal.OkHttpPendingResult.parseResponse(OkHttpPendingResult.java:210)
	at com.google.maps.internal.OkHttpPendingResult.onResponse(OkHttpPendingResult.java:198)
	at com.squareup.okhttp.Call$AsyncCall.execute(Call.java:177)
	at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

   Locked ownable synchronizers:
	- <0x000000066e95b738> (a java.util.concurrent.ThreadPoolExecutor$Worker)

"OkHttp ConnectionPool" #1120 daemon prio=3 os_prio=0 tid=0x00007f3358784800 nid=0x4b9f in Object.wait() [0x00007f33224ef000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:460)
	at com.squareup.okhttp.ConnectionPool$1.run(ConnectionPool.java:101)
	- locked <0x000000065576bf28> (a com.squareup.okhttp.ConnectionPool)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

   Locked ownable synchronizers:
	- <0x000000066e95b828> (a java.util.concurrent.ThreadPoolExecutor$Worker)

**Comment by Ismamai** [comment]
The app was stuck being unable to geocode anything else until we restarted the machine 7 hours later. After reboot, app started geocoding again.

**Comment by Ismamai** [comment]
Added pull request

https://github.com/googlemaps/google-maps-services-java/pull/165

**Comment by Ismamai** [comment]
#165

**Comment by mhubbard** [comment]
@Ismamai I was having a similar issue where the api would keep retrying hit the limit, but block the thread from closing. We were creating a new GeoApiContext for every request which turned out to be the issue, do you create one context then use that for every request? Or do you create a new context for each request? If so try the former.

**Comment by Ismamai** [comment]
We create a GeoApiContext and reuse it for each request. I added the test
case. If you use all your daily usage and try to geocode it will get stuck.

**Comment by domesticmouse** [comment]
Please do not use a new GeoApiContext for each request, it means that the API can't do important things like API request rate limiting.

**Comment by Ismamai** [comment]
As you can see on the Test we attached we reuse the GeoApiContext for all
the requests. We do not create a new GeoApiContext for each request

**Comment by domesticmouse** [comment]
Sorry, I misread your previous statement.

**Comment by domesticmouse** [comment]
I'm currently not seeing this reproduce, even with limit = 2500 and batchSize = 100. By the look of the stack trace, I'm guessing the bug is deep in OkHTTP, which means I should upgrade OkHTTP.

**Comment by Ismamai** [comment]
I was able to reproduce with https://github.com/googlemaps/google-maps-services-java/pull/165

The test first tries to get you to a situation where your request will fail and currently the library retries the request (I think that if it is an OverDailyLimitException, it should just fail and not retry, but this is another issue).

Once you get to the situation where the request will fail, then the library will retry. It is then when the library gets stuck. None of the two callbacks (onResult or onFailire) are called.

I wait for one of the two methods to get called to proceed with my application.

**Comment by ben-manes** [comment]
I've seen this for a few months and had to cancel the request after 10s using a scheduling thread. I then have a nightly job to fix non-geocode records.

However, calling cancel does not release the connection. The library fails to close the response body on an error and, when the pool is eventually consumed, no further requests proceed. This required server restarts to recover from this bad state.

> WARN  [2016-09-14 22:31:24,510] com.squareup.okhttp.OkHttpClient: A connection to https://maps.googleapis.com/ was leaked. Did you forget to close a response body?

As [explained](https://github.com/square/okhttp/issues/2311), `OkHttpPendingResult#getBytes()` must close input stream but fails to do that.

**Comment by Ismamai** [comment]
I did exactly the same, use a timeout to cancel the request if I do not get an answer in the configured time. I was not aware of this other issue that can force you to restart the server.

**Comment by ben-manes** [comment]
@Ismamai Can you re-run your test using my patch? I think that was the root cause.

**Comment by Ismamai** [comment]
@ben-manes do you mean the test I added on[ pull request 165](https://github.com/googlemaps/google-maps-services-java/pull/165) ?
I also have the other test code to check how much time it takes to do geocoding using the library and plain urls. Not sure if you mean the first or the second.

**Comment by ben-manes** [comment]
@Ismamai yes, the first. But any tests you think are good would be nice. Because the connections are not used properly there is also a good chance it reduced the QPS, so the throughput might go up.

**Comment by domesticmouse** [comment]
I've committed @ben-manes' patch, i'd love to know if it fixes this issue. I'll roll a release shortly. Thanks!

**Comment by ben-manes** [comment]
Thanks @domesticmouse!

**Comment by Ismamai** [comment]
Using the last changes and the performanceTest done as result of https://github.com/googlemaps/google-maps-services-java/issues/176 (code in https://github.com/Ismamai/performanceTest)

I get this:

- developent key and QPS=10
    real    0m9.265s
    user    0m0.000s
    sys     0m0.075s

- production key and QPS = 50
    real    0m17.442s
    user    0m0.000s
    sys     0m0.105s
    These time did not make any sense, then I rerun them and I got
    
    real    0m8.500s
    user    0m0.076s
    sys     0m0.106s

- And plain URL
    real    0m7.492s
    user    0m0.015s
    sys     0m0.076s

**Comment by ben-manes** [comment]
Unfortunately the patch did not resolve all of the problems. The client is still able to get into a state where all subsequent requests never complete and timeout. No leak warning is logged, but no requests return to the caller within a 10 second window.

This occurred during a batch import that tried to geocode, resulting in hitting the API limit. Since that used the production key it halted those geocodings. The expected behavior was that it would resolve after QPS block was removed (verified via curl) and the nightly job would correct those entries. Unfortunately that job receives timeouts so the records are not geocoded. As usual, restarting restored functionality.

I think the only short-term resolution is replace this client with a custom one, e.g. using Retrofit.

**Comment by ben-manes** [comment]
I rewrote using Retrofit2 + okhttp3, scaffolding the Gson serializers, and the Java8 call adapters. It passes my tests for geocoding and reverse geocoding, and so far has been more stable in production. Will monitor for a few days.

**Comment by Ismamai** [comment]
@ben-manes I am interested in taking a look at your changes. Not sure where the library gets stuck. I did a change in the library that prevents the system from getting stuck, but it stills does not solve the performance problems that I am getting.

**Comment by domesticmouse** [comment]
@ben-manes do you want to put together a PR? I want to migrate to OkHttp3, and it's going to be a breaking API change when I do it, so we might as well review additional helpful changes.

**Comment by ben-manes** [comment]
I'd think it would be better to discuss the desired end goal, as a PR could expand too far.

I use Retrofit + OkHttp3 heavily, so that would be my preference. I'd lean heavily away from the static API and allow users to configure the connection pool. I also dislike the rate limiting code and sleep logic as error prone. Generally one uses the [token bucket](https://en.wikipedia.org/wiki/Token_bucket) algorithm instead. You might consider adopting a reactive library, e.g. RxJava or Reactor. That should provide rate limiting and batching operators out of the box. I'm torn on whether this should be imposed by the library, though.

Since your current client isn't trying to be dependency free, I think continuing to lean on what's out there is preferable. You aren't blocked by Google approved libraries (e.g. no `thirdparty` upgrade hell) and already try to be mostly serializers and a little glue. Moving more of that glue to external libraries would probably avoid some of these bugs, or at least make it an active project's problem to fix.

**Comment by domesticmouse** [comment]
My elevator pitch desired goal is that this library encapsulate the best practices for people using the Web Services APIs from Java servers, in a production environment. Watching you guys, I'm becoming painfully aware that I'm not running a production service atop this code, so I'm getting a bit blindsided here.

My planned steps are to release the current code base once I can get the tests to stabilise, then figure out how to migrate to OkHttp3. Given how long in the tooth 2.x is, that's a must, but it is also going to break the API. 

I like the idea of using Retrofit to make the code more declarative. I don't know enough about RxJava or Reactor to take that on, I'm not conversant enough with the restrictions that would place on the library users. 

Thoughts?

**Comment by ben-manes** [comment]
I would only use a reactive library if this client is to provide rate limiting and batching natively. If it does then it should try to leverage an existing library. If it defers that to the consumer then returning a `CompletableFuture` is ideal. Since this is server-side Java, I think its okay to make then next major release JDK8-based.

You could also provide two layers. A core layer that uses futures and not magic. A reactive client could be built on top to provide richer functionality. Then users could opt-in to use the high or low level APIs.

**Comment by domesticmouse** [comment]
This looks like a fair chunk of work. I like it as a plan, but it is more work than I currently have time for. So, first step is to migrate to OkHttp3, and then I'll look at the rest. Thanks!

**Comment by unitydynamics** [comment]
@ben-manes would you be willing to share your solution?  I'm having production issues.  Just didn't count on the Java library not being stable.  Google geocoding... Java... these things are not new...

**Comment by ben-manes** [comment]
My solution is mostly in the comment above. There isn't much code, since the provided serialization types are excellent. All I had to do was write a small binding with [Retrofit](http://square.github.io/retrofit/) and migrate my usage. The amount of code is tiny and I haven't had any issues since.

**Comment by unitydynamics** [comment]
I'm new to Retrofit... but wow.  Less code more doing.  Thank you for the response... two very quick questions if you'll oblige.  The "status" and "errorMessage" seem to be magically set... how do I know what the potential status strings are?  Does errorMessage trap exceptions?  I feel like I'm unprepared to handle errors.  

Also... I created a single static version of the service and I'm using it everywhere (then my program terminates).  Do I need to do any kind of cleanup for the service or the okclient? [I just passed in "new OkHttpClient()" to the service constructor]

**Comment by ben-manes** [comment]
Those two fields are deserialized from the response that wraps the results. You would need to look at their Rest api to find out what they could be, if documented. Retrofit will throw an `HttpException` with the http status code (e.g. 400). I treat all geolocation failures as soft errors that are retried in a nightly batch job to correct for missing data.

You could use a rate limiter, e.g. Guava's, if you are concerned about QPS. At the free tier and default paid, I think the default is at a low 5k per minute. If you don't expect much usage, that throttling might be good enough. Otherwise you could use a rate limiter on the client and increase the QPS to handle your workloads.

You shouldn't need to do any cleanup. I use Guice on server and Dagger on Android to avoid statics.

**Comment by domesticmouse** [comment]
I'm going to close this issue out for now. Please raise any remaining work to do as separate issues so I can track. Thanks!
